aspnetcore/security/authentication/cookie.md

```diff
@@ -14,202 +14,332 @@ uid: security/authentication/cookie
+
 ```csharp
-public static class LastChangedValidator
+ValidatePrincipal(CookieValidatePrincipalContext)
+```
+
+An example looks like the following:
+
+```csharp
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.Cookies;
+
+public class CustomCookieAuthenticationEvents : CookieAuthenticationEvents
 {
-    public static async Task ValidateAsync(CookieValidatePrincipalContext context)
+    private readonly IUserRepository _userRepository;
+
+    public CustomCookieAuthenticationEvents(IUserRepository userRepository)
+    {
+        // Get the database from registered DI services.
+        _userRepository = userRepository;
+    }
+
+    public override async Task ValidatePrincipal(CookieValidatePrincipalContext context)
     {
-        // Pull database from registered DI services.
-        var userRepository = context.HttpContext.RequestServices.GetRequiredService<IUserRepository>();
         var userPrincipal = context.Principal;
 
-        // Look for the last changed claim.
-        string lastChanged;
-        lastChanged = (from c in userPrincipal.Claims
-                        where c.Type == "LastUpdated"
-                        select c.Value).FirstOrDefault();
+        // Look for the LastChanged claim.
+        var lastChanged = (from c in userPrincipal.Claims
+                           where c.Type == "LastChanged"
+                           select c.Value).FirstOrDefault();
 
         if (string.IsNullOrEmpty(lastChanged) ||
-            !userRepository.ValidateLastChanged(userPrincipal, lastChanged))
+            !_userRepository.ValidateLastChanged(userPrincipal, lastChanged))
         {
             context.RejectPrincipal();
-            await context.HttpContext.SignOutAsync("MyCookieAuthenticationScheme");
+
+            await context.HttpContext.SignOutAsync(
+                CookieAuthenticationDefaults.AuthenticationScheme);
         }
     }
 }
 ```
 
-This would then be wired up during cookie service registration in the `ConfigureServices` method of *Startup.cs*:
+Register the events instance during cookie service registration in the `ConfigureServices` method. Provide a scoped service registration for your `CustomCookieAuthenticationEvents` class:
 
 ```csharp
-services.AddAuthentication("MyCookieAuthenticationScheme")
-        .AddCookie("MyCookieAuthenticationScheme", options =>
-        {
-            options.Events = new CookieAuthenticationEvents
-            {
-                OnValidatePrincipal = LastChangedValidator.ValidateAsync
-            };
-        });
+services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
+    .AddCookie(options =>
+    {
+        options.EventsType = typeof(CustomCookieAuthenticationEvents);
```

Comment:
I get that we're trying to show off a new 2.0 feature, but the 1.1 way to do this is still there and a lot simpler for the average case. There's no separate doc for the events model right now is there? This belongs somewhere else. @HaoK

Reply:
The 1.1 case conflicts with the whole point of the topic (in spirit): Cookie auth *without Identity* ... i.e., I was kind'a hoping to stay out of that namespace. I also like the new code better. As @PinpointTownes points out, this permits a good DI experience. Devs using something like this without dB access at all seems unlikely.

Reply:
? You're using IUserRepository in both versions...

Reply:
Yes, but in the 1.1 case, it goes with `GetRequiredService`. I just meant that my [@]davidfowl training is that it's a bad pattern. Early on, he corrected something I did and said 'plz don't do that unless you absolutely have to.' lol

Reply:
Ah, proper DI usage is a better argument.

Reply:
Yes, and getting out of the `Microsoft.AspNetCore.Identity` namespace is a *feel good moment* on this topic.

It's kind'a hard to say (in the title of the topic) that this is not using Identity ... when it really *IS*. :smile: ... even if only barely so.